runtime/core/decoder/onnx_asr_model.cc

```diff
@@ -0,0 +1,362 @@
+  // Reset the inner states for new decoding
+  asr_model->Reset();
+  return asr_model;
+}
+
+void OnnxAsrModel::Reset() {
+  offset_ = 0;
+  int required_cache_size;
+  if (num_left_chunks_ > 0) {
+    int required_cache_size = chunk_size_ * num_left_chunks_;
+    offset_ = required_cache_size;
+    att_cache_.resize(num_blocks_ * head_ * required_cache_size *
+                      encoder_output_size_ / head_ * 2);
+    const int64_t att_cache_shape[] = {num_blocks_, head_, required_cache_size,
+                                       encoder_output_size_ / head_ * 2};
+    att_cache_ort_ = Ort::Value::CreateTensor<float>(
+        memory_info_, att_cache_.data(), att_cache_.size(), att_cache_shape, 4);
+
+    att_mask_.resize(required_cache_size + chunk_size_);
+    for (size_t i = 0; i < required_cache_size + chunk_size_; ++i) {
+      att_mask_[i] = i >= required_cache_size;
+    }
+    const int64_t att_mask_shape[] = {1, 1, required_cache_size + chunk_size_};
+    att_mask_ort_ = Ort::Value::CreateTensor<bool>(
+        memory_info_, reinterpret_cast<bool*>(att_mask_.data()),
+        att_mask_.size(), att_mask_shape, 3);
+  } else {
+    att_cache_.resize(0);
+    const int64_t att_cache_shape[] = {num_blocks_, head_, 0,
+                                       encoder_output_size_ / head_ * 2};
+    att_cache_ort_ = Ort::Value::CreateTensor<float>(
+        memory_info_, att_cache_.data(), att_cache_.size(), att_cache_shape, 4);
+
+    att_mask_.resize(0);
+    input_names_.erase(input_names_.begin() + 2);
+    input_names_.pop_back();
+  }
+
+  cnn_cache_.resize(num_blocks_ * encoder_output_size_ *
+                    (cnn_module_kernel_ - 1));
+  const int64_t cnn_cache_shape[] = {num_blocks_, 1, encoder_output_size_,
+                                     cnn_module_kernel_ - 1};
+
+  cnn_cache_ort_ = Ort::Value::CreateTensor<float>(
+      memory_info_, cnn_cache_.data(), cnn_cache_.size(), cnn_cache_shape, 4);
+
+  encoder_outs_.clear();
+}
+
+void OnnxAsrModel::ForwardEncoderFunc(
+    const std::vector<std::vector<float>>& chunk_feats,
+    std::vector<std::vector<float>>* out_prob) {
+  int num_frames = cached_feature_.size() + chunk_feats.size();
+  const int feature_dim = chunk_feats[0].size();
+  int64_t requried_cache_size = chunk_size_ * num_left_chunks_;
+  std::vector<float> model_input;
+  for (size_t i = 0; i < cached_feature_.size(); ++i) {
+    for (int j = 0; j < feature_dim; j++) {
+      model_input.emplace_back(cached_feature_[i][j]);
+    }
+  }
+  for (size_t i = 0; i < chunk_feats.size(); ++i) {
+    for (int j = 0; j < feature_dim; j++) {
+      model_input.emplace_back(chunk_feats[i][j]);
+    }
+  }
+  int num_fill =
+      att_mask_.size() + 1 + right_context_ - subsampling_rate_ - num_frames;
+  if (num_fill > 0) {
+    num_frames += num_fill;
+    for (size_t i = 0; i < num_fill; ++i) {
+      for (int j = 0; j < feature_dim; j++) {
+        model_input.emplace_back(0);
+      }
+    }
+  }
```

Comment:
I find a easy way to handle the last chunk. Just add a slice op in attention.py https://github.com/wenet-e2e/wenet/blob/main/wenet/transformer/attention.py#L90-L94

before:
```py
if mask.size(2) > 0 :  # time2 > 0
            mask = mask.unsqueeze(1).eq(0)  # (batch, 1, *, time2)
            scores = scores.masked_fill(mask, -float('inf'))
            attn = torch.softmax(scores, dim=-1).masked_fill(
                mask, 0.0)  # (batch, head, time1, time2)
```

after:
```py
if mask.size(2) > 0 :  # time2 > 0
            mask = mask.unsqueeze(1).eq(0)  # (batch, 1, *, time2)
            mask = mask[:, :, :, :scores.size(-1)]
            scores = scores.masked_fill(mask, -float('inf'))
            attn = torch.softmax(scores, dim=-1).masked_fill(
                mask, 0.0)  # (batch, head, time1, time2)
```

Reply:
when export onnxmodel, we can keep dynamic axis for `chunk` to allow different input shapes for chunk tensors